### PR TITLE
Drop Result If Too Big

### PR DESCRIPTION
**Pull request by ast2023**
## What changed?
Drop the response from scheduled workflow if the response is too big.

## Why?
If response is too big, the schedule workflow will fail and never restarted.

## How did you test it?
Unit test.

## Potential risks
No.

## Is hotfix candidate?
No.

### CODE REVIEW

File: service/worker/scheduler/activities.go
Comment (dnr) [inline service/worker/scheduler/activities.go:71]: get this from dynamic config: https://github.com/temporalio/temporal/blob/main/common/dynamicconfig/constants.go#L123-L124

File: service/worker/scheduler/activities.go
Comment (dnr) [inline service/worker/scheduler/activities.go:316]: This condition can't happen, the workflow completion would have triggered an error in that case.

The problem is that if the response payload is e.g. (limit – 10 bytes), then the response payload of _this_ activity will be over the limit, because it gets wrapped in WatchWorkflowResponse, and then various local activity marker stuff. So we need to discard it if it's close to the limit. I suggest 1000 bytes as the assumed overhead size.

File: service/worker/scheduler/activities.go
Comment (dnr) [inline service/worker/scheduler/activities.go:331]: have to check Failure size also, failures can exceed the limit

File: service/worker/scheduler/responsebuilder_test.go
Comment (ast2023) [reply service/worker/scheduler/responsebuilder_test.go:335]: My goal is to create easy to read and understand tests. There are more then assertError helper methods in this file. Using testify library does not make tests shorter or more expressive. Usage of if ... { t.Faillf{...} } is easy to understand to any software engineer. It is actually easier to read then require.Equal(a, b, "error message format ", a, b)

File: service/worker/scheduler/activities.go
Comment (ast2023) [inline service/worker/scheduler/activities.go:75]: here is a good example of comment go stale. I updated recordOverheadSize to be 1000 and did not update the comment.

File: service/worker/scheduler/activities.go
Comment (stephanos) [inline service/worker/scheduler/activities.go:390]:
```suggestion
	resp := &schedspb.WatchWorkflowResponse{Status: r.workflowStatus}
	if result != nil {
		resp.ResultFailure = &schedspb.WatchWorkflowResponse_Result{Result: result}
	} else if failure != nil {
		resp.ResultFailure = &schedspb.WatchWorkflowResponse_Failure{Failure: failure}
	}
	return resp
```

Calling it `resp` would make it clearer that it's different from the `result` argument.

File: service/worker/scheduler/activities.go
Comment (stephanos) [inline service/worker/scheduler/activities.go:76]: On first reading, I didn't understand what kind of overhead this is. I now understand that it's the overhead of storing the data, and that we need to deduct it from the storage limit so we don't go over. Could we add a 1-line comment explaining that?

File: service/worker/scheduler/fx.go
Comment (stephanos) [inline service/worker/scheduler/fx.go:1]: I'm noticing that there are three different names that seem to refer to the same thing:
* `maxBlobSize`: the dynamic config property
* `eventStorageSize`: the default for the dynamic config property
* `singleResultStorageSizePerNs`: the copy of the dynamic config property value

And then there's also `resultStorageNumberSize` which seems to be the real limit (since we deducted the record overhead). But from the name it's not clear it's different from `singleResultStorageSizePerNs`.

Would it be reasonable to use similar names to clarify that relationship?

For example `maxEventStorageSizePerNs` for the property, `defaultEventStorageSizePerNs` for the default, `maxEventStorageSizePerNs` again for the copy and `singleResultStorageSize` for the struct?

File: service/worker/scheduler/responsebuilder_test.go
Comment (stephanos) [inline service/worker/scheduler/responsebuilder_test.go:84]: `rb` can be moved to the top of this sub test since it's the same in all nested test cases.

**Review by stephanos — APPROVED** [review]

File: service/worker/scheduler/responsebuilder_test.go
Comment (dnr) [reply service/worker/scheduler/responsebuilder_test.go:325]: It's about consistency: everyone working in this codebase is familiar with testify. Everyone knows to search for "Error Trace" in a long test output because that's the header that testify uses. `require.Equal(a, b)` is really quite easy to read, and that's all you need in 99% of cases. It'll print both values automatically. You can add a contextual message if you want.

Even if custom test helpers are more readable in isolation, they're not in a shared codebase with existing conventions.

**Review by dnr — COMMENTED** [review]
Please do not use auto-merge except to wait for tests to pass _after_ approval. Stephan left a bunch of thoughtful comments and you merged this ignoring them all.

File: service/worker/scheduler/activities.go
Comment (ast2023) [reply service/worker/scheduler/activities.go:74]: This name is making a perfect sense. IMHO BlobSizeLimitError does not make any sence. You would not say it in English. If error needs to be produced once result size crossed eventStorageSize a function `shouldProduceError` or `aboveErrorSizeTreshold`. The fact that we used to use bad names does not mean that we should use bad names. Not sure if we can rename `BlobSizeLimitError` to something more appropriate though.

Comment (ast2023) [reply service/worker/scheduler/activities.go:74]: I am not sure what dynamic config you are referencing. Could you provide more details?

Comment (dnr) [reply service/worker/scheduler/activities.go:74]: You'd read it in English as "blob size limit, at the error level" (there is also a "warn" level).

I'm not saying any name has to read perfectly in English or that BlobSizeLimitError is a great name, I'm saying that this constant has exactly one use in non-test code: the default value for the dynamic config `dynamicconfig.BlobSizeLimitError`.
https://github.com/temporalio/temporal/pull/5369/files#diff-8af44785225f9c30fc88f4265638e6a1f89b2295c7cd609c8f74d5039645d0e3R91
It should be named based on what it is and what it does. Specifically, it does not define a limit on storage size of events, which the current name suggests. It's just a default value, and it's a default value for _blob_ size, which are eventually contained in events but the limit applies to the blob, not the whole event.

Comment (dnr) [reply service/worker/scheduler/activities.go:74]: Also, yes, the fact that we use bad names does mean that we should use bad names, because consistency is more important than the goodness of any specific name. If you want to fix a bad name, please do, but fix it everywhere, not just in one inconsistent corner.

Comment (dnr) [reply service/worker/scheduler/activities.go:74]: See the comment above (and just search for where you used this)

Comment (dnr) [reply service/worker/scheduler/activities.go:74]: I like the second or third. (It's not just "reporting", it fails the workflow.) But I think the name of this constant should include `default`, because the actual error threshold can be changed using dynamic config, and it doesn't always equal this constant.

**Pull request merged**
